added use case template & referenced it in readme (#72)
cncf/cnf-wg: ccdps/NNNN-UC-template.md

# UC-NNNN: Short descriptive title of your use case (TEMPLATE)

> 

Reviewer: The first user story got a glossary, by request.  Should we suggest one?

Author: Agreed. I will add it.

## Involved processes
*TIP: Please mark key processes that are involved in the use case you are presenting. This helps actors to decide if the use case is relevant for them or not. If there are more processes involved please make multiple selection.*
- [ ] Development
- [ ] Deployment
- [ ] System integration
- [ ] Network integration
- [ ] Lifecycle management
- [ ] Operations

## Involved parties
*TIP: In this section you can list main parties who are involved in the use case. For example:<br>

> Business customer of CSP<br>

Reviewer: So the we use 'roles' in agile and that helps a little:

1. Sometimes one person has multiple roles, so it saves the 'jack of all trades does X' kind of user story
2. We don't have to assign roles to companies, for instance, which helps when @jeffsaelens develops his own CNFs despite being an operator

I like your breakdown in general, but I wonder if we can rethink the first list to be roughly around the role list that @taylor was creating - maybe a bit reduced - and fold in the second one.  If we want to talk about 'X buys a CNF from an external party' that is something we could do as a user story by itself, theoretically.

Reviewer 2: Agree with the above comment. May want to use actors as a first-class concept here and cross reference the part where those can be defined.

Reviewer 3: I agree with @claudiobartolini, I think we should define the actors in separate part and solve #54 . This sections can then cross reference that document and we can keep/update them separatley

Author: I was actually having human actors in mind or even personas. I wanted to leave it a bit freestyle to give degree of freedom as we might miss some if we define it as a "foreign key". If you like more to keep actors and cross reference I will change it like that. Otherwise I could go with personas. WDYT?

Reviewer: Whether persona or role, we seem to be settling on the same point; that we should be talking about the hat they're wearing at the time, as opposed to e.g. 'architect employed by a telco' who might wear many hats depending on the telco's internal structure, where they're getting their CNFs from, and so on.

Reviewer 4: I like Ian's hat approach. As we see with focused/targeted calls, a lot of the usual suspects show up as they fill multiple roles. The industry tends to have this trend as well. Its not uncommon to have someone who's been on with a vendor, a telco and on ops and eng throughout their career.

Reviewer: OK, so per the meeting we can have a hat list here with a reasonable list of suggestions and we'll break that out into an independent list with our considered opinions when we've considered @taylor's work more in-depth - that will be a separate commit.

Reviewer 5: The way this is deal with elsewhere, is a single person can fill two roles, but that does not combined the roles, they still exists as distinct roles.

Operations Team of CSP<br>
Service team of CNF Vendor*

## Involved system entities
*TIP: In this section you can list system entities that are involved in the use case. For example:<br>
Data Center Network<br>
Wide Area Network<br>
Cloud Native Platform<br>

> CNF X<br>

Reviewer: I really like this bit.  I would add 'MANO' and 'OSS'.

Author: I have added more examples of entities.

> ## Situation

Reviewer: These two sections (situation/expected behaviour) are probably just one option for telling the story.  It raises the question of 'must you use this template or is it just a helpful place to start?'

Author: Situation is rather to describe point of departure - explaining the scene. Expected behavior is describing how do we expect "the plot" to unfold. If we want to keep comparable structure of use cases I would opt for "must use" template.

Reviewer: I know what you're trying to say, but if you're telling a user story it's usually a stepwise process - 'wants X to happen' then 'does X, expects Y' over and over.  Use cases work out differently because they're more about objects than people (see 'BGP server').  I'm not sure I see how we could make this apply quite right to either of those.

Author: Situation actually tries to capture intents/needs of actors - understanding business/operational context if you will. Expected behavior is more about how, given the intents/needs, of actors the objects should behave.

Reviewer: I suggest we commit with this and how if it works in practice.  Then we will have hard evidence for whether it needs changing rather than personal opinions.

*TIP: Please describe situation related to this use case. Indicate what are intents and expectations of different parties as well as the roles and requirements of different system components. Describe how everything is related. Drawings, flow and system diagrams are beneficial here. This part, as the subject says, shall give the reader clear overview of the situation as the starting point for evaluation of the use case. It should not contain the detailed technical descriptions and analysis.*

## Expected behaviour
*TIP: Here you can provide detailed technical descriptions and depict expected behaviour of system entities and components regardless of cloud nativeness. Please highlight what are different entities expected to provide. Depending on the use case you can look at it from different angles e.g. CNF angle, Cloud Native Platform angle, Network angle etc. This section should give the reader clear understanding of expected behaviour.

> ## Challenges and limitations with Kube-native approach

Reviewer: Seems reasonable.  Misses out the 'this is the platform I am measuring against' that we had in the first user story, but we were talking about whether to extract that (in which case we can reference it here).

# UC-NNNN: Short descriptive title of your use case (TEMPLAT)

## Glossary
*TIP: Please list the key technical terms and acronyms and provide definitions for those.*

## Involved processes

> *TIP: Please mark key processes that are involved in the use case you are presenting. This helps actors to decide if the use case is relevant for them or not. If there are more processes involved please make multiple selection.*

Reviewer 6: Another option can be to uncomment the processes like
```suggestion
> Please uncomment and/or add key processes that are involved in the use case you are presenting. This helps actors to decide if the use case is relevant for them or not.
> - Development
> - Deployment
> - System integration
> - Network integration
> - Lifecycle management
> - Operations
```

- [ ] Development
- [ ] Deployment
- [ ] System integration
- [ ] Network integration
- [ ] Lifecycle management
- [ ] Operations

## Involved parties

> *TIP: In this section you can list main parties who are involved in the use case. For example:<br>

Reviewer 6: ```suggestion
> List main parties involved in the use case. For example:
```

Business customer of CSP<br>
Operations Team of CSP<br>
Service team of CNF Vendor<br>
Etc.*

## Involved system entities

> *TIP: In this section you can list system entities that are involved in the use case. For example:<br>

Reviewer 6: ```suggestion
> List system entities that are involved in the use case. For example:
```

Data Center Network<br>
Wide Area Network<br>
Cloud Native Platform<br>
CNF X<br>
MANO<br>
OSS<br>
Etc.*

## Situation

> *TIP: Please describe situation related to this use case. Indicate what are intents and expectations of different parties as well as the roles and requirements of different system components. Describe how everything is related. Drawings, flow and system diagrams are beneficial here. This part, as the subject says, shall give the reader clear overview of the situation as the starting point for evaluation of the use case. It should not contain the detailed technical descriptions and analysis.*

Reviewer 6: ```suggestion
> Describe situation related to this use case. Indicate what are intents and expectations of different parties as well as the roles and requirements of different system components. Explain how everything is related. Drawings, flow and system diagrams are beneficial here. This part shall give the reader clear overview of the situation as the starting point for evaluation of the use case. It should not contain the detailed technical descriptions and analysis.
```

## Expected behaviour

> *TIP: Here you can provide detailed technical descriptions and depict expected behaviour of system entities and components regardless of cloud nativeness. Please highlight what are different entities expected to provide. Depending on the use case you can look at it from different angles e.g. CNF angle, Cloud Native Platform angle, Network angle etc. This section should give the reader clear understanding of expected behaviour.

Reviewer 6: ```suggestion
behaviour or behavior?
> Provide detailed technical descriptions and depict expected behaviour of system entities and components regardless of cloud nativeness. Highlight what are different entities expected to provide. Depending on the use case you can look at it from different angles e.g. CNF angle, Cloud Native Platform angle, Network angle etc. This section should give the reader clear understanding of expected behaviour.
```

> ## Challenges and limitations with Kube-native approach

Reviewer 7: Is 'Kube-Native' defined somewhere? I thought we were using the term 'Cloud-Native'. Is it the same?

Reviewer: I think by using kube-native, we're saying 'designed to run specifically on Kubernetes'.

<rant>

The face-value definition of 'cloud native' is 'designed to run on a cloud rather than in some other way'. 
 Arguably that might be a virtual machine based cloud, and it certainly doesn't have to be a container based cloud.

You'll note that the CNCF cloud native definition - https://github.com/cncf/toc/blob/main/DEFINITION.md - doesn't actually define cloud native; instead, it sort of says 'you'll know it when you see people using techniques that make it easier to build cloud-native software'.

Since it's a surprisingly woolly term, and people don't consistently mean the same thing when they use it, we should be quite cautious about clarifying what we mean when it comes up.

A fine example of the term being abused is when we talk about a 'cloud native platform' - it doesn't make much sense to refer to cloud platforms as 'cloud native'.

</rant>

Reviewer 6: @iawells I'm totally agree with that distinction but I guess what @rannyh is referring is the fact that Kube-Native has not been documented anywhere in this repo and eventually can be used by other documents.

Reviewer 7: Yeah. I was merely pointing out that the term "kube-native" was not defined anywhere in this document set. Didn't mean to start a debate about the meaning of cloud-nativeness. So going back to my question - are we going to use "kube-native" and add a definition, or use "cloud-native" like in the rest  of this document set?

Author: I have specifically used kube-native there and expected to trigger some discussion with it. I also recall that @taylor was using similar term. Having said that, I believe we need to be focused and true to ourselves and focus on Network Functions that run in Kubernetes cluster + necessary components in and around that cluster.

Cloud native platform or infrastructure is actually real thing. It is infrastructure that exhibits the properties of a cloud and that enables cloud native applications to run on it. Take example of Kubernetes cluster deployed on 10 bare metal nodes on the edge: it can not be called cloud, but it is cloud native infrastructure as it exhibits all properties of cloud (no HA of individual nodes, API driven etc). One more generalized reference for cloud native infrastructure: https://books.google.de/books/about/Cloud_Native_Infrastructure.html?id=1Fk7DwAAQBAJ&printsec=frontcover&source=kp_read_button&redir_esc=y#v=onepage&q&f=false

Reviewer 6: @vukg I'm agree with that so maybe the missing part is the addition of the term into glossary's documentation.

Reviewer: > @iawells I'm totally agree with that distinction but I guess what @rannyh is referring is the fact that Kube-Native has not been documented anywhere in this repo and eventually can be used by other documents.

Agreed.

> *TIP: This section will give us main inputs for discussion and evaluation of best practices. You should describe which challenges and limitations kube-native approach presents for this use case. You can give technical details about what is not possible with kube-native approach and elaborate why is that relevant. You can again look from multiple angles. If there are some established practices that are used to work around these limitations please describe those in the chapter below*

Reviewer 6: ```suggestion
> This section will give us main inputs for discussion and evaluation of best practices. Describe which challenges and limitations kube-native approach presents for this use case. Provide technical details about what is not possible with kube-native approach and elaborate why is that relevant. If there are some established practices that are used to work around these limitations please describe those in the chapter below*
```

### Established practices to overcome challenges and limitations

> *TIP: If such practices exist please elaborate them here in sufficient technical details. Please provide refences. Please give your opinion weather or not these practices can be considered as candidates for CNF best practices in line with [CCDP description](0001-cnf-conformance-definition-proposal-process.md).*

Reviewer 6: ```suggestion
> If such practices exist please elaborate them here in sufficient technical details. Provide references and give your opinion weather or not these practices can be considered as candidates for CNF best practices in line with [CCDP description](0001-cnf-conformance-definition-proposal-process.md).
```

### What needs to be done differently in order to overcome challenges and limitations 

> *TIP: If there are no established practices and you have ideas or if you believe things related to this use case should be done differently please indicate it shortly here. The purpose of this is to earmark some potential inputs for CCDP process. It is not expected to elaborate CCDPs here.

Reviewer 6: ```suggestion
> If there are no established practices and you have ideas or if you believe things related to this use case should be done differently please indicate it shortly here. The purpose of this is to earmark some potential inputs for CCDP process. It is not expected to elaborate CCDPs here.
```
